flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/ChangelogListState.java

```diff
@@ -39,53 +44,83 @@
         extends AbstractChangelogState<K, N, List<V>, InternalListState<K, N, V>>
         implements InternalListState<K, N, V> {
 
-    ChangelogListState(InternalListState<K, N, V> delegatedState) {
-        super(delegatedState);
+    ChangelogListState(
+            InternalListState<K, N, V> delegatedState,
+            KvStateChangeLogger<List<V>, N> changeLogger) {
+        super(delegatedState, changeLogger);
     }
 
     @Override
     public void update(List<V> values) throws Exception {
+        changeLogger.stateUpdated(values, getCurrentNamespace());
         delegatedState.update(values);
     }
 
     @Override
     public void addAll(List<V> values) throws Exception {
+        changeLogger.stateAdded(values, getCurrentNamespace());
         delegatedState.addAll(values);
     }
 
     @Override
     public void updateInternal(List<V> valueToStore) throws Exception {
+        changeLogger.stateUpdated(valueToStore, getCurrentNamespace());
         delegatedState.updateInternal(valueToStore);
     }
 
     @Override
     public void add(V value) throws Exception {
+        if (getValueSerializer() instanceof ListSerializer) {
+            changeLogger.stateElementChanged(
+                    w ->
+                            ((ListSerializer<V>) getValueSerializer())
+                                    .getElementSerializer()
+                                    .serialize(value, w),
+                    getCurrentNamespace());
+        } else {
+            changeLogger.stateAdded(singletonList(value), getCurrentNamespace());
+        }
```

Comment:
Sync up offline.

I am confused with this part.

 if (getValueSerializer() instanceof ListSerializer    is not true,
adding singletonList will also be a problem?

That says when would be the case where getValueSerializer() is not ready
but `changeLogger.stateAdded(singletonList(value), getCurrentNamespace());` can be successful.